[BE] docs: Spring REST Docs와 Swagger UI를 이용한 API 문서 자동화 #205

Byesol opened the pull request:
## 📌 변경 내용 & 이유

기존의 스웨거 기반의 API 문서를 테스트 코드 기반인  Spring REST Docs 방식으로 전환했습니다.

이유
  - 테스트가 통과해야만 문서가 생성되므로 문서의 신뢰성 보장.
  - 비즈니스 코드에 문서 관련 어노테이션 없이 깔끔하게 유지가 가능해짐.

### 주요 변경 사항

  - 모든 컨트롤러에서 `@Tag`, `@Operation` 등 기존 Swagger 어노테이션을 제거
  - `build.gradle`에 `restdocs-api-spec` 및 `swagger.generator` 플러그인을 설정하여 테스트 결과로부터 `openapi3.yaml` 파일을 생성
  - 모든 API의 통합 테스트 코드에 REST Docs를 이용한 문서 생성 로직을 추가
  - 실패 시 반환되는 `ProblemDetail` 형식의 에러 응답까지 문서화
  - 최종적으로 빌드된 Jar 파일에 생성된 Swagger UI가 `static` 리소스로 포함되도록 설정

## 📸 스크린샷 (선택)

<img width="1512" height="1248" alt="image" src="https://github.com/user-attachments/assets/37d22566-0f4d-4299-9659-a7f6732f9743" />


## 🧪 테스트 방법 (선택)

1.  `./gradlew clean test` 명령어로 모든 통합 테스트가 통과하는지 확인
2.  `./gradlew bootJar` 실행 후  브라우저에서 `http://localhost:8080/swagger-ui/index.html` 로 접속하여 REST Docs 기반으로 생성된 API 문서가 정상적으로 보이는지 확인

## 📢 논의하고 싶은 내용



## 🧩 관련 이슈

close #205

<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

* **New Features**
  * Spring REST Docs 및 Swagger UI 기반의 API 문서 자동 생성 및 패키징 기능이 추가되었습니다.

* **Refactor**
  * 기존 Swagger/OpenAPI 어노테이션 및 설정이 모두 제거되고, API 문서화 방식이 전면 교체되었습니다.
  * 일부 관리자용 API가 문서에 노출되도록 변경되었습니다.

* **Tests**
  * 통합 테스트에 API 문서 자동 생성이 연동되어, 테스트 실행 시 API 문서 스니펫이 생성됩니다.
  * 테스트 코드가 REST Docs 기반으로 리팩토링되고, 일부 테스트가 추가/삭제되었습니다.

* **Chores**
  * 불필요한 의존성과 설정이 제거되고, 관련 빌드 설정이 정비되었습니다.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

Byesol committed:
[BE] chore : rest Docs 의존성 추가 및 swagger ui 추가 #205

Byesol committed:
[BE] docs : 북마크 컨트롤러 테스트 rest docs 문서 추가  #205

Byesol committed:
[BE] test: 통합 테스트 rest docs 설정  #205

Byesol committed:
[BE] docs: 테스트용 유틸 클래스 정의  #205

Byesol committed:
[BE] docs: rest docs 사용에 따른 스웨거 설정 제거  #205

Byesol committed:
[BE] docs: responseSchema 추가 및 주석제거  #205

Byesol committed:
[BE] docs: builad.gradle 의 보안 설정 블럭 제거 #205

Byesol committed:
[BE] docs: 카테고리 컨트롤러 테스트 rest docs 문서 추가 #205

Byesol committed:
[BE] docs: FileSource 컨트롤러 테스트에 rest docs 내용 추가 #205

Byesol committed:
[BE] docs: HearitController 테스트에 rest docs 내용 추가 #205

Byesol committed:
[BE] refactor: openapi 파일 생성 주석 처리했던 것을 제거 #205

Byesol committed:
[BE] refactor: IntegrationTest 하나로 통합 #205

Byesol committed:
[BE] docs: AuthControllerTest에 rest docs 내용 추가 #205

Byesol committed:
[BE] refactor: IntegrationTest 하나로 통합됨에 따른 import 구문 수정 #205

Byesol committed:
[BE] docs: KeywordController 에 rest docs 내용 추가 #205

Byesol committed:
[BE] docs: MemberController 에 rest docs 내용 추가 #205

Byesol committed:
[BE] docs: 불필요한 토큰 requestHeader 제거 #205

Byesol committed:
[BE] test: AuthControllerTest에 회원가입 테스트 추가 #205

Byesol committed:
[BE] docs: 불필요한 @Hidden 어노테이션 제거 #205

Byesol committed:
[BE] chore: 불필요한 swaggerUUIFile 설정 제거 #205

Byesol committed:
[BE] Merge remote-tracking branch 'origin/develop-be' into docs/#205-restdocs-swagger

# Conflicts:
#	backend/src/main/java/com/onair/hearit/admin/presentation/AdminCategoryController.java
#	backend/src/main/java/com/onair/hearit/admin/presentation/AdminHearitController.java
#	backend/src/main/java/com/onair/hearit/admin/presentation/AdminKeywordController.java
#	backend/src/main/java/com/onair/hearit/auth/presentation/AuthController.java
#	backend/src/main/java/com/onair/hearit/presentation/BookmarkController.java
#	backend/src/test/java/com/onair/hearit/fixture/IntegrationTest.java
#	backend/src/test/java/com/onair/hearit/presentation/BookmarkControllerTest.java

Byesol committed:
[BE] refactor: 불필요한 swagger 설정 제거 @205

Byesol committed:
[BE] docs: 카테고리별로 그룹화된 히어릿 조회 문서 작성 #205

Byesol committed:
[BE] test: 히어릿 단일 조회 테스트 시 히어릿의 키워드 추가 #205

Byesol committed:
[BE] docs: 로그인 실패 시 401 반환 문서에 추가 #205

Byesol committed:
[BE] refactor: 카테고리 id 존재여부에 따른 예외 검사 추가 #205

Byesol committed:
[BE] docs: PromblemDetails 문서 코드 추가 #205

Byesol committed:
[BE] test: 카테고리 조회 실패 코드 테스트 추가 구현 #205

Byesol committed:
[BE] docs: 북마크컨트롤러 실패 코드 문서 추가  #205

Byesol committed:
[BE] docs: FileSourceControllerTest 실패 코드 문서 추가  #205

Byesol committed:
[BE] docs: HearitControllerTest 실패 코드 문서 추가  #205

Byesol committed:
[BE] docs: KeywordControllerTest 실패 코드 문서 추가  #205

Byesol committed:
[BE] docs: AuthControllerTest 실패 코드 문서 추가  #205

Byesol committed:
[BE] chore: build.gradle 불필요한 주석 제거  #205

Byesol committed:
[BE] refactor: 북마크 생성 실패 문서 identifier 수정  #205

Byesol committed:
[BE] refactor: ApiDocSnippets 패키지명 및 클래스명 변경 #205

Byesol committed:
[BE] chore: SwaggerConfig 완전히 제거 #205

Byesol committed:
[BE] refactor: 북마크 생성 실패 문서 통합 #205

Byesol committed:
[BE] docs: AuthKakaoLoginControllerTest 문서 작성 #205

Byesol committed:
[BE]  refactor: MemberControllerTest 문서 ApiDocSnippets 사용하게 수정 #205

Byesol commented on backend/src/test/java/com/onair/hearit/presentation/MemberControllerTest.java:
```diff
@@ -49,10 +67,29 @@ void getMemberInfo_success() {
     @DisplayName("로그인하지 않은 사용자가 사용자 정보 조회 시, 401 Unauthorized 예외를 반환한다.")
     void getMemberInfo_error_when_isNotLoginedMember() {
-        RestAssured.given().log().all()
+        RestAssured.given(this.spec)
```
수정완료!

yuyoungrhee committed:
[BE] docs: API 문서 full이 아닌 list형식으로 나오도록 수정 #205

yuyoungrhee committed:
[BE] docs: API 문서가 RCUD 순서대로 나오도록 수정 #205

yuyoungrhee committed:
[BE] Merge 'develop-be' into docs/#205-restdocs-swagger #205

# Conflicts:
#	backend/build.gradle

Byesol committed:
[BE] docs: 문서 내 파라미터 default 처리, build gradle 주석 추가 #205

Byesol commented on backend/build.gradle:
```diff
@@ -1,7 +1,15 @@
+buildscript {
+    ext {
+        restdocsApiSpecVersion = '0.18.2'
+    }
+}
+
 plugins {
     id 'java'
     id 'org.springframework.boot' version '3.5.3'
     id 'io.spring.dependency-management' version '1.1.7'
+    id 'com.epages.restdocs-api-spec' version "${restdocsApiSpecVersion}"
+    id 'org.hidetake.swagger.generator' version '2.18.2'
 }
```
가능하던데요

gabean13 commented on backend/build.gradle:
```diff
@@ -1,7 +1,15 @@
+buildscript {
+    ext {
+        restdocsApiSpecVersion = '0.18.2'
+    }
+}
+
```
이렇게 분리해야하는 이유가 있나용?

Byesol replied:
기존에 문서화 적용하며 문제를 해결하는 과정에서 버전 문제인가 싶어서 빠르게 버전을 한 번에 바꾸기 위해 사용했었는데

이제는 굳이 이렇게 할 필요가 없겠네요! 수정했습니다!

Byesol replied:
https://github.com/woowacourse-teams/2025-hEARit/pull/231/commits/7c591badb418af7586929142d2f2c22325b3c4b4

gabean13 commented on backend/build.gradle:
```diff
@@ -60,15 +78,78 @@ dependencies {
 
-tasks.named('jar') {
-    enabled = false
+// GenerateSwaggerUI 태스크가 openapi3 태스크 이후에 실행되도록 설정
+tasks.withType(org.hidetake.gradle.swagger.generator.GenerateSwaggerUI).configureEach {
+    dependsOn 'openapi3'
+}
+
+//  bootJar 빌드 시, 생성된 Swagger UI를 static/swagger-ui에 포함
+tasks.named('bootJar') {
+    dependsOn 'generateSwaggerUISample'
+    from("${tasks.generateSwaggerUISample.outputDir}") {
+        into 'static/swagger-ui'
+    }
+}
+
+tasks.named('generateSwaggerUISample') {
```
이런 내용은 다 build할 때 이뤄져야해서 반드시 build.gradle에 써야하는건가욤??
swagger 관련 설정들을 한 파일에서 관리할 수 있으면 이해하기도 쉽고 관리포인트도 한 곳에 모을 수 잇을 것 같아서 좋을 것 같은데 그건 어려운건가요?

yuyoungrhee replied:
멍PT:

현재 restdocs-api-spec + swagger-generator 조합에서는 build.gradle에 넣는 것이 가장 현실적인 방식입니다.

1. springdoc-openapi와는 다르게,
    - 이 방식은 정적 HTML 생성 기반이라서
    - application.yml이나 SwaggerConfig.java 같은 설정 방식은 아예 작동하지 않음

2. 현재 사용 중인 org.hidetake.swagger.generator 플러그인은
    - Swagger UI HTML/JS를 generateSwaggerUISample에서 자동 생성함
    - index.html도 직접 만들어주기 때문에, 우리가 설정을 주입하거나 후처리할 수 있는 유일한 타이밍이 build.gradle임

yuyoungrhee replied:
방법이 있긴 하네예...

https://www.notion.so/8-2-3-PR-243d39b9c3c3801ca10fc37610e2c632?source=copy_link

Byesol replied:
가콩이 말한 사항에 대해 멍구가 알려준 방법은 일단 이게 머지되고 나중에 이슈파서 시도해봐도 될까요?
build.gradle 설정으로 인해, 이런저런 억까를 많이 당해서 적용하는 게 살짝 무서워서요 ㅠㅠ.

yuyoungrhee replied:
ㅇㅈ합니다

gabean13 commented on backend/src/main/java/com/onair/hearit/auth/presentation/AuthController.java:
```diff
@@ -22,31 +18,30 @@
-                    @ApiResponse(responseCode = "400", description = "중복된 이메일",
-                            content = @Content(schema = @Schema(implementation = ProblemDetail.class)))
-            })
+    //    @Operation(summary = "회원가입", description = "새로운 계정을 생성합니다.",
+//            responses = {
+//                    @ApiResponse(responseCode = "201", description = "회원가입 성공"),
+//                    @ApiResponse(responseCode = "400", description = "중복된 이메일",
+//                            content = @Content(schema = @Schema(implementation = ProblemDetail.class)))
+//            })
+    //TODO 상태코드별 설명 붙이기
```
주석들은 아직 안지우신건가용?

Byesol replied:
엇! 감사합니다. 지웠습니당

gabean13 commented on backend/src/main/java/com/onair/hearit/presentation/MemberController.java:
```diff
@@ -19,17 +13,17 @@
-                    @ApiResponse(responseCode = "401", description = "로그인하지 않은 사용자",
-                            content = @Content(schema = @Schema(implementation = ProblemDetail.class)))
-            })
+    //    @Operation(summary = "사용자 정보 조회", description = "현재 로그인되어있는 사용자의 정보",
+//            responses = {
+//                    @ApiResponse(responseCode = "200", description = "조회 성공"),
+//                    @ApiResponse(responseCode = "401", description = "로그인하지 않은 사용자",
+//                            content = @Content(schema = @Schema(implementation = ProblemDetail.class)))
+//            })
+    //TODO 실패 시PROBLEM DETAIL 내용 문서 작성
```
여기도 주석이 남아있어욥

Byesol replied:
수정했어욥

gabean13 commented on backend/src/test/java/com/onair/hearit/docs/ApiDocSnippets.java:
```diff
@@ -0,0 +1,42 @@
+import org.springframework.restdocs.payload.FieldDescriptor;
+import org.springframework.restdocs.payload.JsonFieldType;
+
+public class ApiDocSnippets {
```
여기서 쓰는건 반복적으로 사용되는 애들을 작성하는건가용? (진짜 몰라서 물어보는 것입니다)

Byesol replied:
네넹. api 문서를 위해 반환하는 dto에 관한 문서를 작성해야하는데

PagedResponse, 와 ProblemDetails 의 경우 여러 api에서 반환하고 있어서 따로 뺏어용.

그치만 테스트를 미리 작성하는 건 아니고, 문서만 작성하는 거기 때문에  fixture 디렉토리가 아닌 docs 폴더에 만들었습니당.

gabean13 commented on backend/src/test/java/com/onair/hearit/fixture/IntegrationTest.java:
```diff
@@ -1,16 +1,24 @@
 @Sql("/dbclean.sql")
 @ActiveProfiles("integration-test")
 @SpringBootTest(webEnvironment = SpringBootTest.WebEnvironment.RANDOM_PORT)
+@ExtendWith(RestDocumentationExtension.class)
```
어노테이션 순서 코드 컨벤션 맞춰주세용

Byesol replied:
수정했습니다!

gabean13 reviewed (commented):
와우 벡터 테스트코드 작성하느라 고생하셨어요..

yuyoungrhee reviewed (changes requested):
벡터 가능하다면 같은방식으로 default parameter 처리 해줄 수 있나요

얘네 두 개 빠졌습니다!
- /api/v1/categories
- /api/v1/keywords/recommend

나머지는 괜찮은 것 같습니다 넘 수고하셨음

Byesol committed:
[BE] refactor: 어노테이션 컨벤션에 따른 정렬 순서 변경 #205

Byesol committed:
[BE] refactor: 사용하지 않는 ApiDocSnippets 제거 #205

Byesol committed:
[BE] refactor: 버전을 직접 명시하도록 수정 #205

Byesol committed:
[BE] refactor: 사용하지 않는 주석 제거 #205

Byesol committed:
[BE] refactor: 문서 내 파라미터 default 값 추가 #205

Byesol committed:
[BE] Merge remote-tracking branch 'origin/docs/#205-restdocs-swagger' into docs/#205-restdocs-swagger

Byesol commented:
멍구 https://github.com/woowacourse-teams/2025-hEARit/pull/231/commits/842d8f9fc3024184aae31f6a22d9d7a0059ba8b2 에서 default parameter 추가했습니다!

JO-eusan commented on backend/src/main/java/com/onair/hearit/admin/presentation/AdminKeywordController.java:
```diff
@@ -5,7 +5,7 @@
 import com.onair.hearit.dto.request.PagingRequest;
 import com.onair.hearit.dto.response.PagedResponse;
-import io.swagger.v3.oas.annotations.Hidden;
+import java.net.URI;
```
안쓰는 import가 있네용

Byesol replied:
엥 분명 인텔리제이 설정으로 안 쓰는 import문 지우기를 해놨는데 남아있네영. 감사합니다!

수정했어요!
https://github.com/woowacourse-teams/2025-hEARit/pull/231/commits/78d89b5a1bc556d8fe38afa97659b2796790dee7

JO-eusan commented on backend/src/main/java/com/onair/hearit/admin/presentation/AdminHearitController.java:
```diff
@@ -6,7 +6,7 @@
 import com.onair.hearit.dto.request.PagingRequest;
 import com.onair.hearit.dto.response.PagedResponse;
-import io.swagger.v3.oas.annotations.Hidden;
+import java.net.URI;
```
여기도 안쓰는 import가 있어용

Byesol replied:
https://github.com/woowacourse-teams/2025-hEARit/pull/231/commits/78d89b5a1bc556d8fe38afa97659b2796790dee7

JO-eusan commented on backend/src/main/java/com/onair/hearit/presentation/BookmarkController.java:
```diff
@@ -6,11 +6,7 @@
-import io.swagger.v3.oas.annotations.media.Schema;
-import io.swagger.v3.oas.annotations.responses.ApiResponse;
-import io.swagger.v3.oas.annotations.tags.Tag;
+import java.net.URI;
```
전반적으로 컨트롤러에 불필요한 import가 있네유

yuyoungrhee reviewed (dismissed):
확인했습니다! build.gradle 설정 분리는 좀 복잡해보여서 나중에 해보죠!!

JO-eusan reviewed (commented):
벡터 코멘트 달았어요! import 문만 제거해주시면 될듯요
다음주에 build.gradle 관해서 이야기 해봅시당

Byesol committed:
[BE] refactor: 쓰지 않는 import 문 제거 #205

Byesol committed:
Merge remote-tracking branch 'origin/develop-be' into docs/#205-restdocs-swagger

# Conflicts:
#	backend/src/main/java/com/onair/hearit/admin/presentation/AdminCategoryController.java
#	backend/src/main/java/com/onair/hearit/admin/presentation/AdminKeywordController.java
#	backend/src/main/java/com/onair/hearit/application/CategoryService.java
#	backend/src/main/java/com/onair/hearit/auth/presentation/AuthController.java
#	backend/src/main/java/com/onair/hearit/presentation/CategoryController.java
#	backend/src/test/java/com/onair/hearit/auth/presentation/AuthControllerTest.java
#	backend/src/test/java/com/onair/hearit/auth/presentation/AuthKakaoLoginControllerTest.java
#	backend/src/test/java/com/onair/hearit/presentation/CategoryControllerTest.java

Byesol committed:
[BE] chore: refresh 토큰의 만료 시간 변경 #205

Byesol committed:
[BE] refactor: 사용하지 않는 import 제거 #205

Byesol committed:
[BE] refactor: 사용하지 swagger 어노테이션 제거 #205

Byesol committed:
[BE] refactor: refresh 토큰 추가에 따른 api 문서 내용 추가 #205

Byesol committed:
[BE] refactor: url 주소 변경에 따른 api 문서 내용 변경 #205

Byesol committed:
[BE] refactor: 필요없어진 404 테스트 제거 #205

Byesol committed:
[BE] docs: 스웨거 문서 서버에 개발 서버 추가  #205

Byesol committed:
[BE] Merge remote-tracking branch 'origin/develop-be' into docs/#205-restdocs-swagger

# Conflicts:
#	backend/src/test/resources/application-integration-test.properties

Byesol committed:
[BE] : 랜덤 히어릿 keyword 추가에 따른 테스트 문서 수정 #205

Byesol committed:
[BE] docs: AuthControllerTest 내 문서의 요청 스키마 수정 #205

JO-eusan commented on backend/src/main/java/com/onair/hearit/application/CategoryService.java:
```diff
@@ -1,5 +1,6 @@
 package com.onair.hearit.application;
 
+import com.onair.hearit.common.exception.custom.NotFoundException;
```
안쓰는 import가 있네용

Byesol replied:
죄송해요.. 인텔리제이 자동 제거에 너무 의존했습니다. 앞으로 좀 더 꼼꼼하게 보고  pr 날릴께요. 제거했습니다.

Byesol replied:
https://github.com/woowacourse-teams/2025-hEARit/pull/231/commits/68ef8419a8d63f4a6a3b6e62ca75df2c8239b9f1

JO-eusan commented on backend/src/main/java/com/onair/hearit/presentation/KeywordController.java:
```diff
@@ -32,14 +28,12 @@ public ResponseEntity<List<KeywordResponse>> readKeywords(
     }
 
-    @Operation(summary = "오늘의 추천 키워드 조회", description = "오늘의 추천 키워드를 조회합니다.")
     @GetMapping("/recommend")
```
이 api를 지금 쓰고 있는 곳이 있을까요?
제가 알기로는 없어진 것 같은데 추후 확장될 가능성이 거의 없어보여서 지워도 되지 않을까요?

Byesol replied:
제거완료했습니다
https://github.com/woowacourse-teams/2025-hEARit/pull/231/commits/c8cd459772e9af4d32d95440237b3f7abd52505f

JO-eusan commented on backend/src/main/java/com/onair/hearit/presentation/MemberController.java:
```diff
@@ -19,17 +13,10 @@
-                            content = @Content(schema = @Schema(implementation = ProblemDetail.class)))
-            })
+    
     @GetMapping("/me")
```
갑자기 드는 의문인데 `/members/me`가 무슨 api인가요?

yuyoungrhee replied:
로그인된 사용자정보(자신의 정보)를 호출하는 api입니다..
members/me가 일반적인 표현이라고해서 따랐습니다! current랑 고민했는데 보통 me를 쓴다 하더라고요

JO-eusan reviewed (commented):
테스트 코드 문서로 작성하느라 고생하셨어요..벡터
간단한 코멘트 남겨둘게요! 확인부탁드려요

Byesol committed:
[BE] refactor: 사용하지 않는 import 제거 #205

Byesol committed:
[BE] refactor: 사용하지 않는 오늘의 키워드 기능 제거 #205

yuyoungrhee reviewed (approved):
확인했습니다!!! 수고많으셨음 Approve할게요

gabean13 reviewed (approved):
고생햇음둥
